pandas/core/indexes/interval.py

```diff
@@ -1090,6 +1091,36 @@ def equals(self, other):
     def overlaps(self, other):
         return self._data.overlaps(other)
 
+    def intersection(self, other, sort=False):
+        other = self._as_like_interval_index(other)
+
+        # GH 19016: ensure set op will not return a prohibited dtype
```

Comment:
I it worth factoring out the this check from here and _setop to a helper routine

Reply:
May I add the decorator for the check when more setops in IntervalIndex's own form is implemented?

Reply:
if you can do it here would be great; this is pretty common code